### PR TITLE
User policy enforcement

### PR DESCRIPTION
**Pull request by matajoh**
Support for user enforcement in policy.

### CODE REVIEW

File: internal/guest/runtime/hcsv2/uvm.go [hunk 1]
```diff
@@ -292,6 +294,65 @@ func setupSandboxHugePageMountsPath(id string) error {
...
+	users, err := user.ParsePasswdFileFilter(filepath.Join(spec.Root.Path, "/etc/passwd"), func(user user.User) bool {
+		return uint32(user.Uid) == uid
+	})
+	if err != nil {
+		return "", err
+	}
+
+	if len(users) != 1 {
+		return "", errors.Errorf("expected exactly 1 user matched '%d'", len(users))
+	}
```
Comment (SeanTAllen) [inline internal/guest/runtime/hcsv2/uvm.go:307]: excellent this one right here.

File: internal/guest/runtime/hcsv2/uvm.go [hunk 2]
```diff
@@ -292,6 +294,65 @@ func setupSandboxHugePageMountsPath(id string) error {
...
+				return securitypolicy.IDName{}, nil, 0, err
+			}
+			groups = append(groups, securitypolicy.IDName{ID: strconv.FormatUint(uint64(gid), 10), Name: groupName})
+		}
+	}
+
+	umask := syscall.Umask(0022)
+	if spec.Process.User.Umask != nil {
+		umask = int(*spec.Process.User.Umask)
+	}
```
Comment (SeanTAllen) [inline internal/guest/runtime/hcsv2/uvm.go:351]: what's going on here?

we are setting 0022 as the default? Why 0022? I'm sure its a perfectly reasonable thing. I think that the why of that would be good to have in a comment explaining what this function does (explain all but the why of the default umask in particular should be in)

File: internal/guest/runtime/hcsv2/uvm.go [hunk 3]
```diff
@@ -648,13 +717,22 @@ func (h *Host) ExecProcess(ctx context.Context, containerID string, params prot.
...
 	if params.IsExternal || containerID == UVMContainerID {
 		var allowStdioAccess bool
 		envToKeep, allowStdioAccess, err = h.securityPolicyEnforcer.EnforceExecExternalProcessPolicy(
 			params.CommandArgs,
 			processParamEnvToOCIEnv(params.Environment),
 			params.WorkingDirectory,
 			params.OCIProcess.NoNewPrivileges,
+			userName,
+			groupNames,
+			umask,
```
Comment (SeanTAllen) [inline internal/guest/runtime/hcsv2/uvm.go:735]: This doesnt go through runc, are these values actually used? From my understanding for how exec external works, I don't think we care about this. Ditto no new privs. Have a look. I don't think we should be enforcing these 4 for exec external.

File: pkg/securitypolicy/framework_objects.json [hunk 4]
```diff
@@ -87,6 +87,22 @@
         "no_new_privileges": {
             "introduced_version": "0.2.0",
             "default_value": true
+        },
+        "user": {
+            "introduced_version": "0.2.1",
+            "default_value": {
+                "umask": 0022,
```
Comment (SeanTAllen) [inline pkg/securitypolicy/framework_objects.json:94]: shouldnt this be a string? "0022"

File: pkg/securitypolicy/framework.rego [hunk 5]
```diff
@@ -1142,3 +1171,33 @@ errors["invalid noNewPrivileges"] {
...
+    some process in data.external_processes
+    command_ok(process.command)
+    workingDirectory_ok(process.working_dir)
+    user_ok(process.user)
+}
+
+errors["invalid user"] {
+    input.rule in ["create_container", "exec_in_container", "exec_external"]
+    not user_matches
+}
```
Comment (SeanTAllen) [inline pkg/securitypolicy/framework.rego:1203]: newline at end?

File: internal/guest/runtime/hcsv2/uvm.go (hunk 2 above)
Comment (matajoh) [reply internal/guest/runtime/hcsv2/uvm.go:351]: Great point. 0022 is the default used by `runc`:

https://github.com/opencontainers/runc/blob/2430a98e646a408003d1c327578ce602f6580cd4/libcontainer/rootfs_linux.go#L149

(provided I'm reading that codebase correctly) so I figured it was a safe bet to use it here. Not good to leave it undocumented, though.

File: internal/guest/runtime/hcsv2/uvm.go (hunk 3 above)
Comment (matajoh) [reply internal/guest/runtime/hcsv2/uvm.go:735]: agreed as stated

File: pkg/securitypolicy/framework_objects.json (hunk 4 above)
Comment (matajoh) [reply pkg/securitypolicy/framework_objects.json:94]: I'm unclear on exactly how Go's umask interface works.

Comment (matajoh) [reply pkg/securitypolicy/framework_objects.json:94]: Yes, this should be a string.

File: internal/guest/runtime/hcsv2/uvm.go (hunk 1 above)
Comment (matajoh) [reply internal/guest/runtime/hcsv2/uvm.go:307]: thanks! I'd love to take credit but this is copy pasta from where this check happens elsewhere. Super important.

File: internal/guest/runtime/hcsv2/uvm.go (hunk 3 above)
Comment (matajoh) [reply internal/guest/runtime/hcsv2/uvm.go]: this is done in latest.

File: pkg/securitypolicy/framework_objects.json (hunk 4 above)
Comment (matajoh) [reply pkg/securitypolicy/framework_objects.json]: Changed.

File: pkg/securitypolicy/framework.rego (hunk 5 above)
Comment (matajoh) [reply pkg/securitypolicy/framework.rego]: Done.

**Pull request closed without merging**